src/Workspaces/Core/Portable/FindSymbols/SymbolTree/SymbolTreeInfo_Metadata.cs

```diff
@@ -85,194 +98,482 @@ private static Metadata GetMetadataNoThrow(PortableExecutableReference reference
+                }
+            }
 
-                // We don't include internals from metadata assemblies.  It's less likely that
-                // a project would have IVT to it and so it helps us save on memory.  It also
-                // means we can avoid loading lots and lots of obfuscated code in the case the
-                // dll was obfuscated.
-                if (IsPublic(type.Attributes))
+            private void AddTypeDefinitionNameParts(
+                TypeDefinitionHandle handle, List<string> simpleNames)
+            {
+                var typeDefinition = _metadataReader.GetTypeDefinition(handle);
+                var declaringType = typeDefinition.GetDeclaringType();
+                if (declaringType.IsNil)
                 {
-                    var definition = MetadataDefinition.Create(reader, type);
-                    definitionMap.Add(definition.Name, definition);
+                    // Not a nested type, just add the containing namespace.
+                    AddNamespaceParts(typeDefinition.NamespaceDefinition, simpleNames);
                 }
+                else
+                {
+                    // We're a nested type, recurse and add the type we're declared in.
+                    // It will handle adding the namespace properly.
+                    AddTypeDefinitionNameParts(declaringType, simpleNames);
+                }
+
+                // Now add the simple name of the type itself.
+                simpleNames.Add(GetMetadataNameWithoutBackticks(_metadataReader, typeDefinition.Name));
             }
-        }
 
-        private static void LookupMetadataDefinitions(
-            MetadataReader reader, NamespaceDefinition namespaceDefinition,
-            OrderPreservingMultiDictionary<string, MetadataDefinition> definitionMap)
-        {
-            foreach (var child in namespaceDefinition.NamespaceDefinitions)
+            private void AddNamespaceParts(
+                NamespaceDefinitionHandle namespaceHandle, List<string> simpleNames)
+            {
+                if (namespaceHandle.IsNil)
+                {
+                    return;
+                }
+
+                var namespaceDefinition = _metadataReader.GetNamespaceDefinition(namespaceHandle);
+                AddNamespaceParts(namespaceDefinition.Parent, simpleNames);
+                simpleNames.Add(_metadataReader.GetString(namespaceDefinition.Name));
+            }
+
+            private void AddTypeReferenceNameParts(TypeReferenceHandle handle, List<string> simpleNames)
             {
-                var definition = MetadataDefinition.Create(reader, child);
-                definitionMap.Add(definition.Name, definition);
+                var typeReference = _metadataReader.GetTypeReference(handle);
+                var namespaceString = _metadataReader.GetString(typeReference.Namespace);
+
+                // NOTE(cyrusn): Unfortunately, we are forced to allocate here
```

Comment:
>NOTE(cyrusn) [](start = 19, length = 12)

File the issue on CoreFX and link it here so that we know to update the code when the issues is fixed.